### PR TITLE
Fix possible race in caml_mem_map on Windows

### PR DESCRIPTION
**Pull request by dra27**
It appears I managed to leave some `printf`-debugging in my 2018 hacking from the concurrent minor collector in `platform.c`. The first commit in this PR nukes those statements - they were never intended to see the light of day! 🙈

As noted in https://github.com/ocaml/ocaml/pull/10831#discussion_r780850386, there is a possibility of a race when trimming the memory block.

I'm not certain about the implementation of the loop here, but I'll open this as a starter. In particular, I think there may be a "fast path" being missed. If the start address is aligned as required, it's almost certainly better to avoid the risk of a race and allow a tiny waste of memory at the end of the block. I don't know the details of Go's allocator, but I can't see how we'd end up with enough (any?) parallel contention to warrant a retry loop counter, especially as in all the various test runs of this in the last 4 years, the "error" traps there before weren't triggered (in the testsuite, etc.).

### CODE REVIEW

**Comment by gadmm** [comment]
Thanks for looking into it. Could the second call to VirtualAlloc not fail for other reasons (e.g. out of memory)? In this case we need the loop counter.
I am not sure that this function is (too) performance-sensitive; you'd need a benchmark before deciding it is worth looking into optimisations.

**Comment by dra27** [comment]
If we're out of memory then the first `VirtualAlloc` should also return `NULL`, terminating the loop (but I freely confess to not having yet given it sufficient thought or checking for merging!).

**Comment by gadmm** [comment]
From the documentation, MEM_COMMIT charges the requested memory against the memory limit (RAM + paging files) whereas MEM_RESERVE does not. As I understand, the second could fail due to OOM when the first one succeeds.

**Review by gadmm — COMMENTED** [review]
I am fairly confident that this is a good solution since it is used in Go: https://github.com/golang/go/blob/c27a3592aec5f46ae18f7fd3d9ba18e69e2f1dcb/src/runtime/malloc.go#L813-L828.

To summarise the discussion, I think it is possible for the first VirtualAlloc to repeatedly succeed while the second one repeatedly fails, for instance in out-of-memory situations, so you need a retry counter to exit the loop.

**Comment by gasche** [comment]
@gadmm do you approve this change? (The printf-removal is obvious enough, but the loop...)

**Comment by gadmm** [comment]
As I just explained, it just needs a retry counter like in the Go implementation. For the loop I trust the quality of Go's engineering solving the exact same issue, so I am ready to approve it after that.

**Comment by gasche** [comment]
Thanks! It was not clear to me if your message was approving of the current PR or suggesting a change (or confirming a previous change suggestion, etc.)

**Comment by dra27** [comment]
An overdue return to this PR. I have:
- Removed `caml_mem_unmap` because the Windows implementation is incorrect and fundamentally not possible. The function is only used in the Unix implementation of `caml_mem_map`, so the easiest thing seems to be to inline it.
- Added a guard to `caml_mem_decommit` to avoid calling `VirtualFree` for a 0-bytes block. If my analysis is correct, the `printf` started being seen by @xavierleroy on Inria's CI after #11046 was merged, which I think is correct as that PR (correctly) causes domains to attempt to decommit 0 bytes when a domain is created.
- Tweaked `caml_mem_map` to retry for _exactly_ the correct condition - exhausting either available memory and exhausting available address space return distinct error codes from `VirtualAlloc`, so this loop retries only when the reservations are being contended - one reservation will always succeed, so the loop should always terminate. That race has been seen a few times on Jenkins.

This PR is being tested along with a systhreads PR, appropriately, in [precheck#666](https://ci.inria.fr/ocaml/job/precheck/666/)

**Comment by gasche** [comment]
We use `caml_mem_unmap` in #11082: the contiguous area for "minor heaps" is mmaped, and when we need a new, bigger area we unmap the previous one and we map a new one.

I guess we could *not* unmap under Windows; in Linux terms, this would consume more virtual address space, which is not desirable, but it may work fine in practice.

**Comment by dra27** [comment]
Thanks, @gasche - let me have a look. This should be fine - the issue is that you can't unmap _part_ of the section, which I don't think you're doing there, so that'd be OK.

**Comment by gadmm** [comment]
Not a Windows programmer. Is it standard to use `GetLastError` in a programmatic way? The VirtualAlloc page does not list the error values, and e.g. https://docs.microsoft.com/en-us/windows/win32/api/errhandlingapi/nf-errhandlingapi-getlasterror says
> The error codes returned by a function are not part of the Windows API specification [...]

**Comment by dra27** [comment]
It is normal to use `GetLastError` that way, yes - and, alas, not documenting the actual error codes actually used by functions is possibly the most irritating thing about MSDN!

**Comment by dra27** [comment]
I've added some additional commits which restore `caml_mem_unmap` - I've also added some debugging assertions to be sure that both `munmap` and `VirtualFree` are returning the correct statuses.

It is important that `caml_mem_unmap` is only used to deallocate entire regions i.e. the `size` given should be the same as the `size` originally passed to `caml_mem_map`. I have added for the debug runtime a possible piece of tracking code to check that. I have also rebased this on to #11082 and ensured that the test added in that PR does not trip the assertions added here when run with the debug runtime.

File: runtime/platform.c
Comment (gadmm) [inline runtime/platform.c]: How about something like `VirtualFree(...) && CAMLassert(false);` and `munmap(...) || CAMLassert(false);` (instead of new macros)?

File: runtime/platform.c
Comment (gadmm) [inline runtime/platform.c:168]: My understanding is that `caml_mem_map` is called the first time before any domains is spawned, so no lock is necessary. (I think there might be other easy ways to avoid the lock, but let me know if this is enough for you.)

**Review by gadmm — APPROVED** [review]
It globally looks good, and the debugging code to ensure proper `VirtualAlloc` usage is a very nice touch.

Some simplifications are proposed.

Something to think about in the future is merging the debug `mmap_block` business with the freeing of all mmaped blocks at shutdown in the "pooled" mode (not yet implemented IIRC).

File: runtime/platform.c
Comment (dra27) [reply runtime/platform.c:168]: Hah - I had considered this, but worried about its being brittle 🙂 Maybe on this occasion I was being overly cautious - it's in the debug runtime, and the lock is essentially pointless at the moment, so perhaps with more comments it would be fine to note that it isn't racy.

Comment (gadmm) [reply runtime/platform.c:168]: I thought about brittleness. Let's simply `CAMLassert(caml_domain_alone())`, this should be enough.

File: runtime/platform.c
Comment (gadmm) [reply runtime/platform.c]: Plus, I find the macros confusing: usually CAMLassert means something which is not executed when assertions are turned off.

File: runtime/platform.c
Comment (dra27) [reply runtime/platform.c:168]: `caml_domain_alone` doesn't suffice in its current form - at the first call, there are _no_ domains. In the fixup I've pushed, I opted to check `caml_num_domains_running` directly - an alternative would be to relax the definition of `caml_domain_alone` to be `<= 1` rather than `== 1`, but that felt riskier (if not even wrong) in other places.

Comment (dra27) [reply runtime/platform.c:168]: (but thanks for the suggestion, as it certainly controls any brittleness!)

File: runtime/platform.c
Comment (dra27) [reply runtime/platform.c]: What I did was ugly and unnecessary - thank you for objecting! The compiler is perfectly clever enough to generate exactly the same code for:

```c
if (!VirtualFree(...))
  ((void) 0)
```
and
```c
VirtualFree(...)
```
so I re-wrote them with the more sensible ifs which should have been there in the first place!

**Comment by dra27** [comment]
Thanks for the review, @gadmm! Do those two additional commits look OK?

File: runtime/platform.c
Comment (gadmm) [reply runtime/platform.c:168]: Thanks, this looks fine to me for debug code (I'd have `<= 1` instead of `== 0` but whatever).

**Comment by gadmm** [comment]
Yes, it looks ok. Let's clean-up history and ask @gasche to double-stamp.

File: runtime/platform.c
Comment (dra27) [reply runtime/platform.c:168]: I changed that when squashing!

**Comment by dra27** [comment]
History tidied!

**Review by gasche — APPROVED** [review]
Approved on @gadmm behalf.

Note: you track mmap in a skip list for debug reasons. But some other parts of the runtime had to track mmaped sizes for other reasons as well:

https://github.com/ocaml/ocaml/blob/832153cf74c26b1ea48808b655188104b8d54088/runtime/caml/fiber.h#L59

I wonder if we should consider (later) this technique of using a skip list to store mmap-area metadata for more things than just Windows debug logic.

**Comment by dra27** [comment]
> I wonder if we should consider (later) this technique of using a skip list to store mmap-area metadata for more things than just Windows debug logic.

Quite possibly! A small addition note (for the proverbial record) - the debug logic is not Windows-specific, that's in all both flavours of the debug runtime to try to ensure that we never use `caml_mem_unmap` to trim memory blocks.

**Pull request merged**
